### PR TITLE
CB-11576: Fix FreeIPA race condition when adding a role

### PR DESCRIPTION
**Pull request by jamisonbennett**
Fix FreeIPA race condition when adding a role. It was possible for the
FreeIPA management service to fail to add a host if multiple hosts were added at the same time with the same role assignment.

This was manually tested in a debugger by forcing the failure
condition to occur. Both hosts were successfully added.

See detailed description in the commit message.

### CODE REVIEW

**Comment by christmasferret** [comment]
nice catch. keep up the good work! ahahahah

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java [hunk 1]
```diff
@@ -53,7 +53,18 @@ public void addRoleAndPrivileges(Optional<Service> service, Optional<Host> host,
         if (roleRequest != null && StringUtils.isNotBlank(roleRequest.getRoleName())) {
             Set<Role> allRole = ipaClient.findAllRole();
             Optional<Role> optionalRole = allRole.stream().filter(role -> role.getCn().equals(roleRequest.getRoleName())).findFirst();
-            Role role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            Role role;
```
Comment (akanto) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java:56]: logic is getting more and more complex here, we shall add a bit more test coverage, focusing also on the edge cases

Comment (jamisonbennett) [reply freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java:56]: I added some tests for the race condition and also for some of the prior corner cases that were previously insufficiently unit tested.

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/model/Host.java [hunk 2]
```diff
@@ -81,4 +81,17 @@ public Boolean getHasKeytab() {
     public void setMemberOfRole(List<String> memberOfRole) {
         this.memberOfRole = memberOfRole;
     }
+
+    @Override
+    public String toString() {
+        return "Host{"
```
Comment (aakanser) [inline freeipa-client/src/main/java/com/sequenceiq/freeipa/client/model/Host.java:87]: You could also use `String.format("....", dn, krbprincipal,...)`.

You could have a `public static char/String` for the apostrophe instead of retyping.

Comment (aakanser) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/model/Host.java:87]: maybe even a helper method `apostropheMe(string)` that just lets you do `apostropheMe(dn)`.

**Review by aakanser — APPROVED** [review]
looks good to me, left one style comment

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/model/Host.java (hunk 2 above)
Comment (jamisonbennett) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/model/Host.java:87]: There are many other files in cloudbreak which just use the `'`. So I think I am going to leave it as is, at least for now.

**Comment by jamisonbennett** [comment]
It appears that the jenkins jobs are failing to pull maven dependencies right now. I can't seem to get them to pass, so I am going to just give it some time, and hopefully, it starts working soon.

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java [hunk 3]
```diff
@@ -53,7 +53,18 @@ public void addRoleAndPrivileges(Optional<Service> service, Optional<Host> host,
...
+            try {
+                role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            } catch (FreeIpaClientException e) {
+                if (!FreeIpaClientExceptionUtil.isDuplicateEntryException(e)) {
+                    LOGGER.error("Failed to add the role [{}]", roleRequest.getRoleName(), e);
+                    throw e;
+                }
+                LOGGER.debug("The role [{}] was recently created in a different thread, retrieve it", roleRequest.getRoleName(), e);
+                allRole = ipaClient.findAllRole();
+                role = allRole.stream().filter(r -> r.getCn().equals(roleRequest.getRoleName())).findFirst().get();
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java:66]: this looks like a duplicate of line 54-55. Could you move this to a common method?

**Review by lacikaaa — CHANGES_REQUESTED** [review]
thanks for the change! good catch!

**Review by lacikaaa — APPROVED** [review]
please resolve conflicts
